[IMP] l10n_au: tax report headings

richard-willdooit opened the pull request:
- Provide consistent formatting of totals - amounts owed to the ATO and
amounts owed by the ATO are now formatted the same.

- Also, the titles are NOT subtotals of the rows below, leading to
misleading sales / purchase totals (which are not transferred to
BAS in any case).

Description of the issue/feature this PR addresses:

Inconsistent formatting of BAS Tax Report Headings / Totals

Current behavior before PR:

Headings had meaningless unreportable subtotal figures
Total lines were formatted inconsistently

Desired behavior after PR is merged:

Remove unneeded figures and standardise 2 subtotals



--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

jco-odoo commented:
@richard-willdooit  The CLA should not be needed here anymore.  Can you remove that commit, so we can merge this?

richard-willdooit commented:
@jco-odoo Thanks.  Done.

jco-odoo commented:
@richard-willdooit Checking https://www.ato.gov.au/assets/0/104/694/815/bc3893a5-0d8c-4ed6-84f4-6af8fbcb9d12.pdf I don't know why we make the G9 line invisible.  Is not it more logical to show it?  (maybe you did it to give a right amount on the header, where we removed the amount anyways)

richard-willdooit commented:
@jco-odoo I did not make it invisible - I took away the link to the header.

Now it prints as a bold line (with the correct figure) as the "bottom line" for the first section.

It is actually now consistent with G20, which is the "bottom line" for the second section.

richard-willdooit commented:
@jco-odoo @william-andre Bump....

william-andre commented on addons/l10n_au/data/account_tax_report_data.xml:
```diff
@@ -80,14 +81,15 @@
         <field name="sequence" eval="10"/>
         <field name="formula">((G1-(G2+G3+G4))+G7)/11</field>
-        <field name="parent_id" ref="account_tax_report_gstrpt_sale_total"/>
+        <field name="parent_id" eval="False"/>
```
I'm not sure why you want to change this if you are using `formula=None` for `account_tax_report_gstrpt_sale_total` anyways.

This will just remove this line from the report, is that what you want?

richard-willdooit replied:
@william-andre 

Changing this leaves it on the report, but in a bold attribute, not as a sub attribute.  This increases the consistency in the report between sales and purchases section.

With this change, the visual layout more closely resembles the layout from the Australian Taxation Office.

We use this mod already, directly with our clients.

william-andre replied:
> Changing this leaves it on the report, but in a bold attribute, not as a sub attribute. 

My bad, I was confused because that behavior has been changed in newer versions.

________________
> This increases the consistency in the report between sales and purchases section.

What do you mean?
https://github.com/odoo/odoo/blob/d6ff8c49d7a0775c4744dfecdeb679595b791e76/addons/l10n_au/data/account_tax_report_data.xml#L182-L188

> With this change, the visual layout more closely resembles the layout from the Australian Taxation Office.

I couldn't find what you are referring to, I only found this
![image](https://aws1.discourse-cdn.com/business5/uploads/manager1/original/3X/d/9/d9dd065de035e23b31edb0f7f07b58347330eb22.png)
![image](https://docplayer.net/docs-images/50/15476020/images/page_7.jpg)
And what seems to be the official document: https://www.ato.gov.au/assets/0/104/694/815/68f311eb-9d6a-4367-8ea5-3dbb52fb859e.pdf

richard-willdooit replied:
@william-andre Essentially, the 3 figures that are important are G9, which gets transferred to 1A, and G20, which gets transferred to 1B.

In the existing report, G20 is printed on the Odoo report without a parent id, and stands out nice and cleanly, but G9 is lost in the top section.

Essentially, the report makes it look like a subtotal, but it is not - it is another figure / 11 - to be transferred,  When we were queried about it, we checked with a few of our customers and the (small) sample we took all liked the visual improvement and consistency of changing G9 to parallel G20.

richard-willdooit replied:
@william-andre Ahh, wait a sec, I see with the parent Id - I need to check my local code - hang on...

richard-willdooit replied:
@william-andre Sorry, was a while since I looked at this.  Check account_tax_report_gstrpt_g20b, which is the sum of g20a and gst_only_purchases.... That is what is printing as "G20"

richard-willdooit replied:
![Screenshot from 2022-01-13 18-39-45](https://user-images.githubusercontent.com/1011100/149381343-710494f9-1a74-4acb-95cb-00f090cf7724.png)

william-andre replied:
Looking at what seems to be the official document: https://www.ato.gov.au/assets/0/104/694/815/68f311eb-9d6a-4367-8ea5-3dbb52fb859e.pdf, it looks like that it the mistake, and not the opposite.

richard-willdooit replied:
I see what you are saying - but in the end, the 2 figures that have to be extracted are G9 and G20 - and they do not stand out in the current format.

IF we were to introduce, as hinted, a "Transfer to 1A / 1B" or "Summary" as hinted by the main screenshots, I would agree.
![Screenshot from 2022-01-14 09-38-50](https://user-images.githubusercontent.com/1011100/149482223-31258967-bafc-4307-8453-9ede353f9f19.png)

![Screenshot from 2022-01-14 09-38-38](https://user-images.githubusercontent.com/1011100/149482218-7955644c-2165-4088-a398-fc1afdab024d.png)

william-andre replied:
Adding these 1A and 1B make more sense to me than hacking the hierarchy, and keeps things more standard.
Every single other software I saw does it like that.

richard-willdooit commented:
@william-andre OK, so I have reverted the subtotal in sales, moved the subtotal in purchases, made a new summary section, and also cleaned up the comparison section (which was wrong with a "tag")

william-andre commented on addons/l10n_au/data/account_tax_report_data.xml:
```diff
@@ -190,25 +193,73 @@
+    <!-- Summary -->
+    <record id="account_tax_report_gstrpt_summay" model="account.tax.report.line">
+        <field name="name">Summary</field>
+        <field name="report_id" ref="tax_report"/>
```
`ValueError: External ID not found in the system: l10n_au.tax_report`

richard-willdooit commented:
@william-andre Seems to be passing all checks for v13 now

william-andre commented on addons/l10n_au/data/account_tax_report_data.xml:
```diff
@@ -78,6 +79,7 @@
 
     <record id="account_tax_report_gstrpt_g9" model="account.tax.report.line">
         <field name="name">G9: GST on sales (G8 divided by eleven)</field>
+        <field name="code">G9</field>
```
This is only supported since version 13.0

https://github.com/odoo/enterprise/commit/5d7ed69e423d6e6814a5a185279afd3b634c531b

richard-willdooit replied:
Oh dear, sorry - but when it was green on all tests?  And this is v13?

OK, will check it again

william-andre replied:
> richard-willdooit wants to merge 1 commit into odoo:13.0 from WilldooIT:13.0-l10n_au-tax-headings

Yes :)

Regarding the tests, the localization were very poorly tested until version 14.0, and really tested only since version 15.0. Sorry about that, but everything merged before version 15.0 needs to be tested manually.

richard-willdooit replied:
Well, I hope this one nails it now.  I created a v13 db and tested it.

william-andre reviewed (commented):
```
  File "/home/odoo/git/odoo/enterprise/account_reports/models/account_report.py", line 1175, in get_report_informations
    'main_html': self.get_html(options),
  File "/home/odoo/git/odoo/enterprise/account_reports/models/account_report.py", line 1213, in get_html
    lines = self._get_lines(options, line_id=line_id)
  File "/home/odoo/git/odoo/enterprise/account_reports/models/account_generic_tax_report.py", line 417, in _get_lines
    return self._get_lines_by_grid(options, line_id, data)
  File "/home/odoo/git/odoo/enterprise/account_reports/models/account_generic_tax_report.py", line 459, in _get_lines_by_grid
    lines[index] = self._build_total_line(total_line, balances_by_code, hierarchy_level, total_period_number, options)
  File "/home/odoo/git/odoo/enterprise/account_reports/models/account_generic_tax_report.py", line 573, in _build_total_line
    period_total = safe_eval(report_line.formula, eval_dict)
  File "/home/odoo/git/odoo/odoo/odoo/tools/convert.py", line 32, in <lambda>
    safe_eval = lambda expr, ctx={}: s_eval(expr, ctx, nocopy=True)
  File "/home/odoo/git/odoo/odoo/odoo/tools/safe_eval.py", line 383, in safe_eval
    pycompat.reraise(ValueError, ValueError('%s: "%s" while evaluating\n%r' % (ustr(type(e)), ustr(e), expr)), exc_info[2])
  File "/home/odoo/git/odoo/odoo/odoo/tools/pycompat.py", line 13, in reraise
    raise value.with_traceback(tb)
  File "/home/odoo/git/odoo/odoo/odoo/tools/safe_eval.py", line 360, in safe_eval
    return unsafe_eval(c, globals_dict, locals_dict)
  File "", line 1, in <module>
ValueError: <class 'NameError'>: "name 'G9' is not defined" while evaluating
'G9'
```

william-andre reviewed (approved):
Thanks a lot!

@robodoo r+
